### PR TITLE
feat(iast): overhead control engine (OCE) skeleton

### PR DESCRIPTION
**Pull request by avara1986**
## Description
Basic structure to implement the features of Overhead control engine (OCE)
OCE control the number of requests to analyze, number of vulnerabilities to report and number of vulnerabilities to analyze. 

## Reviewer Checklist
- [ ] Title is accurate.
- [ ] Description motivates each change.
- [ ] No unnecessary changes were introduced in this PR.
- [ ] PR cannot be broken up into smaller PRs.
- [ ] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.
- [ ] Add to milestone.

### CODE REVIEW

**Review by gnufede — APPROVED** [review]

**Review by juanjux — APPROVED** [review]

**Review by juanjux — DISMISSED** [review]

File: ddtrace/appsec/iast/__init__.py [hunk 1]
```diff
@@ -7,7 +7,33 @@
...
-the end of the request
-"""
+In order to have the better performance, the Overhead control engine (OCE) helps us to control the overhead of our
+wrapped functions. We should create a class that inherit from `ddtrace.appsec.iast.taint_sinks._base.VulnerabilityBase`
+and register with `ddtrace.appsec.iast.oce`.
+
+@oce.register
+class MyVulnerability(VulnerabilityBase):
+    vulnerability_type = "MyVulnerability"
+    evicence_type = "kind_of_Vulnerability"
```
Comment (gnufede) [inline ddtrace/appsec/iast/__init__.py:17]:
```suggestion
    evidence_type = "kind_of_Vulnerability"
```

File: ddtrace/appsec/iast/overhead_control_engine.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    from typing import Set
+    from typing import Type
+
+MAX_REQUESTS = int(os.environ.get("DD_IAST_MAX_CONCURRENT_REQUESTS", 2))
+MAX_VULNERABILITIES_PER_REQUEST = int(os.environ.get("DD_IAST_VULNERABILITIES_PER_REQUEST", 2))
+
+
+class Operation(object):
+    """Common operation related to Overhead Control Engine (OCE). Every vulnerabilities/taint_sinks should inherit
+    from this class. OCE instance requests for these methods to control the overhead in each request.
```
Comment (gnufede) [inline ddtrace/appsec/iast/overhead_control_engine.py:21]:
```suggestion
    from this class. OCE instance calls these methods to control the overhead produced in each request.
```

File: ddtrace/appsec/iast/overhead_control_engine.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    _quota = MAX_VULNERABILITIES_PER_REQUEST
+
+    @classmethod
+    def reset(cls):
+        cls._quota = MAX_VULNERABILITIES_PER_REQUEST
+
+    @classmethod
+    def decrement_quota(cls):
+        if cls._quota > 0:
+            cls._quota -= 1
```
Comment (gnufede) [inline ddtrace/appsec/iast/overhead_control_engine.py:34]: I believe this should also be controlled with a lock.

How about a single function `acquire_quota` that checks for `has_quota` and `decrements_quota` if there is quota available?

Comment (gnufede) [reply ddtrace/appsec/iast/overhead_control_engine.py:34]: Also: I insist in renaming this `quota` to `vulnerability_quota` and `request` to `request_quota`

File: ddtrace/appsec/iast/taint_sinks/_base.py [hunk 4]
```diff
@@ -1,28 +1,88 @@
...
+        """Build a IastSpanReporter instance to report it in the `AppSecIastSpanProcessor` as a string JSON
+
+        TODO: check deduplications if DD_IAST_DEDUPLICATION_ENABLED is true
+        """
+        if cls.has_quota():
+
+            span = tracer.current_root_span()
+            if not span:
+                log.debug("No root span in the current execution. Skipping IAST taint sink.")
+                return None
```
Comment (gnufede) [inline ddtrace/appsec/iast/taint_sinks/_base.py:57]: I'd move this check outside of `has_quota` to lower the pressure over the threading lock

**Review by juanjux — APPROVED** [review]

**Review by juanjux — APPROVED** [review]

File: ddtrace/appsec/iast/taint_sinks/_base.py [hunk 5]
```diff
@@ -49,14 +49,12 @@ def report(cls, evidence_value=""):
 
         TODO: check deduplications if DD_IAST_DEDUPLICATION_ENABLED is true
         """
-        if cls.has_quota():
-
+        if cls.acquire_quota():
             span = tracer.current_root_span()
             if not span:
                 log.debug("No root span in the current execution. Skipping IAST taint sink.")
                 return None
```
Comment (gnufede) [inline ddtrace/appsec/iast/taint_sinks/_base.py:56]: I'd move cls.aquire_quota to after this check, to avoid aquiring_quota for nothing

**Review by gnufede — APPROVED** [review]

**Pull request merged**